README.rst
############
Pyramid LDAP
############

``pyramid_ldap`` provides LDAP authentication services for your Pyramid
application.  Thanks to the ever-awesome `SurveyMonkey
<http://surveymonkey.com>`_ for sponsoring the development of this package!

See the documentation at
https://docs.pylonsproject.org/projects/pyramid_ldap/en/latest/ for more
information.

This package will only work with Pyramid 1.3 and later.

Installation
------------

``pyramid_ldap`` uses ``pyldap`` which in turn requires ``libldap2`` and
``libsasl2`` development headers installed.

On Ubuntu 16.04 you can install them using the command ``apt-get install libldap2-dev libsasl2-dev``.



CHANGES.txt
0.3.1.post1 (2018-05-15)
------------------------

- version changed in setup.py

0.3.1 (2018-05-15)
------------------

- Added wheels license to setup.cfg

0.3 (2018-05-15)
----------------

- Python 3 is now supported! This is possible thanks to python-ldap supporting
  python 3

- BREAKING CHANGE: pyramid_ldap now depends on python-ldap >= 3.0, which
  supports python 2 and 3. If you upgrade pyramid_ldap without upgrading
  python-ldap, you will see failures where python-ldap receives text arguments
  instead of bytes. See below for more information.

- python-ldap >= 3.0 stopped accepting ``bytes`` type arguments for many
  ``LDAPObject`` methods. We now use the ``pyramid.compat.text_`` function to
  support text arguments for python 2 and 3.

- In versions 0.2 and lower, invalid LDAP strings passed to
  ``pyramid_ldap.Connector.authenticate`` would raise ``ldap.FILTER_ERROR``. In
  0.3 and up we use ``ldap.filter.escape_filter_chars`` to properly escape these
  prior to running the LDAP search. Applications using pyramid_ldap have always
  been responsible for making sure the ``login`` value fits the
  ``ldap.login_filter_tpl`` setting in their application. However, if you relied
  on ``ldap.FILTER_ERROR`` to catch bad username formats (such as
  ``CORP\username``, where the unescaped ``\`` is disallowed in LDAP searches),
  note that now ``authenticate`` will return ``None`` instead.

- Use tox for testing against Python 2.7, 3.6, PEP8, coverage and building
  docs. Setup Travis CI to run tox. Use pylons-sphinx-themes to fix broken
  Read The Docs builds.
  See: https://github.com/Pylons/pyramid_ldap/pull/22

0.2
---

- Prevent the use of zero-length password authentication.
  See: https://github.com/Pylons/pyramid_ldap/pull/13

0.1
---

-  Initial version


pyramid_ldap/tests.py
import contextlib
import unittest
import sys

from pyramid.compat import (
    text_type,
    text_,
    )
from pyramid import testing
from pyramid.exceptions import ConfigurationError


class Test_includeme(unittest.TestCase):
    def _callFUT(self, config):
        from pyramid_ldap import includeme
        includeme(config)

    def test_it(self):
        config = DummyConfig()
        self._callFUT(config)
        self.assertEqual(config.directives,
                         ['ldap_setup', 'ldap_set_login_query',
                          'ldap_set_groups_query'])


class Test__ldap_decode(unittest.TestCase):
    def _callFUT(self, val):
        from pyramid_ldap import _ldap_decode
        return _ldap_decode(val)

    def test_decode_str(self):
        result = self._callFUT('abc')
        self.assertEqual(type(result), text_type)
        self.assertEqual(result, text_('abc'))

    def test_decode_list(self):
        result = self._callFUT(['abc', 'def'])
        self.assertEqual(type(result), list)
        self.assertEqual(result[0], text_('abc'))
        self.assertEqual(result[1], text_('def'))

    def test_decode_tuple(self):
        result = self._callFUT(('abc', 'def'))
        self.assertEqual(type(result), tuple)
        self.assertEqual(result[0], text_('abc'))
        self.assertEqual(result[1], text_('def'))

    def test_decode_dict(self):
        import ldap
        result = self._callFUT({'abc': 'def'})
        self.assertTrue(isinstance(result, ldap.cidict.cidict))
        self.assertEqual(result[text_('abc')], text_('def'))

    def test_decode_nested(self):
        import ldap
        result = self._callFUT({'abc': ['def', 'jkl']})
        self.assertTrue(isinstance(result, ldap.cidict.cidict))
        self.assertEqual(result[text_('abc')], [text_('def'), text_('jkl')])

    def test_undecodeable(self):
        uid = b'\xdd\xafw:PuUO\x8a#\x17\xaa\xc2\xc7\x8e\xf6'
        result = self._callFUT(uid)
        self.assertTrue(isinstance(result, bytes))


class Test_groupfinder(unittest.TestCase):
    def _callFUT(self, dn, request):
        from pyramid_ldap import groupfinder
        return groupfinder(dn, request)

    def test_no_group_list(self):
        request = testing.DummyRequest()
        request.ldap_connector = DummyLDAPConnector('dn', None)
        result = self._callFUT('dn', request)
        self.assertEqual(result, None)

    def test_with_group_list(self):
        request = testing.DummyRequest()
        request.ldap_connector = DummyLDAPConnector('dn', [('groupdn', None)])
        result = self._callFUT('dn', request)
        self.assertEqual(result, ['groupdn'])


class Test_get_ldap_connector(unittest.TestCase):
    def _callFUT(self, request):
        from pyramid_ldap import get_ldap_connector
        return get_ldap_connector(request)

    def test_no_connector(self):
        request = testing.DummyRequest()
        self.assertRaises(ConfigurationError, self._callFUT, request)

    def test_with_connector(self):
        request = testing.DummyRequest()
        request.ldap_connector = True
        result = self._callFUT(request)
        self.assertEqual(result, True)


class Test_ldap_setup(unittest.TestCase):
    def _callFUT(self, config, uri, **kw):
        from pyramid_ldap import ldap_setup
        return ldap_setup(config, uri, **kw)

    def test_it_defaults(self):
        from pyramid_ldap import Connector
        config = DummyConfig()
        self._callFUT(config, 'ldap://')
        self.assertEqual(config.prop_name, 'ldap_connector')
        self.assertEqual(config.prop_reify, True)
        request = testing.DummyRequest()
        self.assertEqual(config.prop(request).__class__, Connector)


class Test_ldap_set_groups_query(unittest.TestCase):
    def _callFUT(self, config, base_dn, filter_tmpl, **kw):
        from pyramid_ldap import ldap_set_groups_query
        return ldap_set_groups_query(config, base_dn, filter_tmpl, **kw)

    def test_it_defaults(self):
        import ldap
        config = DummyConfig()
        self._callFUT(config, 'dn', 'tmpl')
        self.assertEqual(config.registry.ldap_groups_query.base_dn, 'dn')
        self.assertEqual(config.registry.ldap_groups_query.filter_tmpl, 'tmpl')
        self.assertEqual(config.registry.ldap_groups_query.scope,
                         ldap.SCOPE_SUBTREE)
        self.assertEqual(config.registry.ldap_groups_query.cache_period, 0)


class Test_ldap_set_login_query(unittest.TestCase):
    def _callFUT(self, config, base_dn, filter_tmpl, **kw):
        from pyramid_ldap import ldap_set_login_query
        return ldap_set_login_query(config, base_dn, filter_tmpl, **kw)

    def test_it_defaults(self):
        import ldap
        config = DummyConfig()
        self._callFUT(config, 'dn', 'tmpl')
        self.assertEqual(config.registry.ldap_login_query.base_dn, 'dn')
        self.assertEqual(config.registry.ldap_login_query.filter_tmpl, 'tmpl')
        self.assertEqual(config.registry.ldap_login_query.scope,
                         ldap.SCOPE_ONELEVEL)
        self.assertEqual(config.registry.ldap_login_query.cache_period, 0)


class TestConnector(unittest.TestCase):
    def _makeOne(self, registry, manager):
        from pyramid_ldap import Connector
        return Connector(registry, manager)

    def test_authenticate_no_ldap_login_query(self):
        manager = DummyManager()
        inst = self._makeOne(None, manager)
        self.assertRaises(ConfigurationError, inst.authenticate,
                          'login', 'password')

    def test_authenticate_search_none_login_returns_none(self):
        manager = DummyManager()
        registry = Dummy()
        registry.ldap_login_query = DummySearch([('a', 'b')])
        inst = self._makeOne(registry, manager)
        self.assertEqual(inst.authenticate(None, 'password'), None)

    def test_authenticate_search_none_password_returns_none(self):
        manager = DummyManager()
        registry = Dummy()
        registry.ldap_login_query = DummySearch([('a', 'b')])
        inst = self._makeOne(registry, manager)
        self.assertEqual(inst.authenticate('login', None), None)

    def test_authenticate_search_returns_none_result(self):
        manager = DummyManager()
        registry = Dummy()
        registry.ldap_login_query = DummySearch([])
        inst = self._makeOne(registry, manager)
        self.assertEqual(inst.authenticate('login', 'password'), None)

    def test_authenticate_empty_password(self):
        manager = DummyManager()
        registry = Dummy()
        registry.ldap_login_query = DummySearch([('a', 'b')])
        inst = self._makeOne(registry, manager)
        self.assertEqual(inst.authenticate('login', ''), None)

    def test_authenticate_search_returns_one_result(self):
        manager = DummyManager()
        registry = Dummy()
        registry.ldap_login_query = DummySearch([('a', 'b')])
        inst = self._makeOne(registry, manager)
        self.assertEqual(inst.authenticate('login', 'password'), ('a', 'b'))

    def test_authenticate_search_bind_raises(self):
        import ldap
        manager = DummyManager([None, ldap.LDAPError])
        registry = Dummy()
        registry.ldap_login_query = DummySearch([('a', 'b')])
        inst = self._makeOne(registry, manager)
        self.assertEqual(inst.authenticate('login', 'password'), None)

    def test_authenticate_escapes_login(self):
        manager = DummyManager()
        registry = Dummy()
        registry.ldap_login_query = DummySearch([('a', 'b')])
        inst = self._makeOne(registry, manager)
        inst.authenticate('BAD\login', 'password')
        expected_escaped_login = 'BAD\\5clogin'
        self.assertEqual(registry.ldap_login_query.kw['login'],
                         expected_escaped_login)

    def test_authenticate_escapes_password(self):
        manager = DummyManager()
        registry = Dummy()
        registry.ldap_login_query = DummySearch([('a', 'b')])
        inst = self._makeOne(registry, manager)
        inst.authenticate('login', 'bad\*()password')
        expected_escaped_password = r'bad\5c\2a\28\29password'
        self.assertEqual(registry.ldap_login_query.kw['password'],
                         expected_escaped_password)

    def test_user_groups_no_ldap_groups_query(self):
        manager = DummyManager()
        inst = self._makeOne(None, manager)
        self.assertRaises(ConfigurationError, inst.user_groups, None)

    def test_user_groups_search_returns_result(self):
        manager = DummyManager()
        registry = Dummy()
        registry.ldap_groups_query = DummySearch([('a', 'b')])
        inst = self._makeOne(registry, manager)
        self.assertEqual(inst.user_groups(None), [('a', 'b')])

    def test_user_groups_execute_raises(self):
        import ldap
        manager = DummyManager()
        registry = Dummy()
        registry.ldap_groups_query = DummySearch([('a', 'b')], ldap.LDAPError)
        inst = self._makeOne(registry, manager)
        self.assertEqual(inst.user_groups(None), None)


class Test_LDAPQuery(unittest.TestCase):
    def _makeOne(self, base_dn, filter_tmpl, scope, cache_period):
        from pyramid_ldap import _LDAPQuery
        return _LDAPQuery(base_dn, filter_tmpl, scope, cache_period)

    def test_query_cache_no_rollover(self):
        inst = self._makeOne(None, None, None, 1)
        inst.last_timeslice = sys.maxsize
        inst.cache['foo'] = 'bar'
        self.assertEqual(inst.query_cache('foo'), 'bar')

    def test_query_cache_with_rollover(self):
        inst = self._makeOne(None, None, None, 1)
        inst.cache['foo'] = 'bar'
        self.assertEqual(inst.query_cache('foo'), None)
        self.assertEqual(inst.cache, {})
        self.assertNotEqual(inst.last_timeslice, 0)

    def test_execute_no_cache_period(self):
        inst = self._makeOne('%(login)s', '%(login)s', None, 0)
        conn = DummyConnection('abc')
        result = inst.execute(conn, login='foo')
        self.assertEqual(result, 'abc')
        self.assertEqual(conn.arg, ('foo', None, 'foo'))

    def test_execute_with_cache_period_miss(self):
        inst = self._makeOne('%(login)s', '%(login)s', None, 1)
        conn = DummyConnection('abc')
        result = inst.execute(conn, login='foo')
        self.assertEqual(result, 'abc')
        self.assertEqual(conn.arg, ('foo', None, 'foo'))

    def test_execute_with_cache_period_hit(self):
        inst = self._makeOne('%(login)s', '%(login)s', None, 1)
        inst.last_timeslice = sys.maxsize
        inst.cache[('foo', None, 'foo')] = 'def'
        conn = DummyConnection('abc')
        result = inst.execute(conn, login='foo')
        self.assertEqual(result, 'def')


class DummyLDAPConnector(object):
    def __init__(self, dn, group_list):
        self.dn = dn
        self.group_list = group_list

    def user_groups(self, dn):
        return self.group_list


class Dummy(object):
    def __init__(self, *arg, **kw):
        pass


class DummyConfig(object):
    introspectable = Dummy

    def __init__(self):
        self.registry = Dummy()
        self.directives = []

    def add_directive(self, name, fn):
        self.directives.append(name)

    def set_request_property(self, prop, name, reify=False):
        self.prop_reify = reify
        self.prop_name = name
        self.prop = prop

    def action(self, discriminator, callable, introspectables=()):
        if callable:
            callable()


class DummyManager(object):
    def __init__(self, with_errors=()):
        self.with_errors = with_errors

    @contextlib.contextmanager
    def connection(self, username=None, password=None):
        yield self
        if self.with_errors:
            e = self.with_errors.pop(0)
            if e is not None:
                raise e


class DummySearch(object):
    def __init__(self, result, exc=None):
        self.result = result
        self.exc = exc

    def execute(self, conn, **kw):
        if self.exc is not None:
            raise self.exc
        self.kw = kw
        return self.result


class DummyConnection(object):
    def __init__(self, result):
        self.result = result

    def search_s(self, *arg):
        self.arg = arg
        return self.result


sampleapp/views.py
import pprint

from pyramid.view import view_config, forbidden_view_config
from pyramid.response import Response
from pyramid.httpexceptions import HTTPFound
from pyramid.security import remember, forget, authenticated_userid, unauthenticated_userid

from pyramid_ldap import get_ldap_connector

@view_config(route_name='sampleapp.root', permission='view', renderer='templates/root.pt')
def logged_in(request):
    return dict(dn=authenticated_userid(request))

@view_config(route_name='sampleapp.logout')
def logout(request):
    headers = forget(request)
    return HTTPFound('/', headers=headers)
    # # We could as well keep the user in a clean logged out page
    # # by doing 
    # return Response('Logged out', headers=headers)

@view_config(route_name='sampleapp.login', renderer='templates/login.pt')
@forbidden_view_config(renderer='templates/login.pt')
def login(request):
    url = request.current_route_url()
    login = ''
    password = ''
    error = ''

    if 'form.submitted' in request.POST:
        login = request.POST['login']
        password = request.POST['password']
        connector = get_ldap_connector(request)
        data = connector.authenticate(login, password)
        if data is not None:
            pprint.pprint(data)
            dn = data[0]
            headers = remember(request, dn)
            return HTTPFound('/', headers=headers)
        else:
            error = 'Invalid credentials'
            
    return dict(
        login_url=url,
        login=login,
        password=password,
        error=error,
        )
                


pyramid_ldap/declarative.py
import ldap
from pyramid.settings import asbool

def scope_converter(v):
    if 'subtree' in v.lower():
        return ldap.SCOPE_SUBTREE
    if 'onelevel' in v.lower():
        return ldap.SCOPE_ONELEVEL
    if 'base' in v.lower():
        return ldap.SCOPE_BASE
    raise ValueError('Unknown scope %s' % v)

def search_args_from_settings(settings, prefix, default_scope):
    def get_setting(name):
        return settings.get(prefix + name)

    search_args = []

    for key, converter, default in (
        ('%sbase_dn' % prefix, None, None), 
        ('%sfilter_tmpl' % prefix, None, None),
        ('%sscope' % prefix, scope_converter, default_scope),
        ):
            setting = get_setting(key)
            if setting is None:
                setting = default
            if setting is None:
                raise ValueError('Must specify %s for an LDAP search' % key)
            if converter is not None:
                setting = converter(setting)
            search_args.append(setting)
    return search_args

def setup_from_settings(config, settings, prefix='pyramid_ldap.'):

    def get_setting(name):
        return settings.get(prefix + name)

    pool_args = {}
    uri = get_setting('uri')
    pool_args['uri'] = uri
    if uri is None:
        raise ValueError('If you include pyramid_ldap, you must set a '
                         '%suri configuration key' % prefix)
    for name, converter in (
        ('bind', None), ('passwd', None), ('size', int), 
        ('retry_max', int), ('retry_delay', float), ('use_tls', asbool),
        ('timeout', int), ('use_pool', asbool),
        ):
        setting = get_setting(name)
        if setting is not None:
            if converter is not None:
                setting = converter(setting)
            pool_args[name] = setting

    config.ldap_setup_pool(**pool_args)
    login_search_args = search_args_from_settings(
        settings,
        prefix + 'login_search.',
        ldap.SCOPE_BASE
        )
    config.ldap_set_login_search(*login_search_args)
    groups_search_args = search_args_from_settings(
        settings,
        prefix + 'groups_search.',
        ldap.SCOPE_SUBTREE
        )
    config.ldap_set_groups_search(*groups_search_args)

def includeme(config):
    config.add_directive('setup_ldap_from_settings', setup_from_settings)


pyramid_ldap/__init__.py
try:
    import ldap
except ImportError:  # pragma: no cover
    # this is for benefit of being able to build the docs on rtd.org
    class ldap(object):
        LDAPError = Exception
        SCOPE_ONELEVEL = None
        SCOPE_SUBTREE = None


import logging
import pprint

import time

from ldap.filter import escape_filter_chars

from pyramid.exceptions import ConfigurationError
from pyramid.compat import (
    iteritems_,
    text_,
)


try:
    from ldappool import ConnectionManager

except ImportError:  # pragma: no cover
    class ConnectionManager(object):
        def __init__(self, *arg, **kw):
            # this is for benefit of being able to build the docs on rtd.org.
            raise ImportError('No module named ldappool')


logger = logging.getLogger(__name__)


class _LDAPQuery(object):
    """ Represents an LDAP query.  Provides rudimentary in-RAM caching of
    query results."""
    def __init__(self, base_dn, filter_tmpl, scope, cache_period):
        self.base_dn = base_dn
        self.filter_tmpl = filter_tmpl
        self.scope = scope
        self.cache_period = cache_period
        self.last_timeslice = 0
        self.cache = {}

    def __str__(self):
        return ('base_dn=%(base_dn)s, filter_tmpl=%(filter_tmpl)s, '
                'scope=%(scope)s, cache_period=%(cache_period)s' %
                self.__dict__)

    def query_cache(self, cache_key):
        result = None
        now = time.time()
        ts = _timeslice(self.cache_period, now)

        if ts > self.last_timeslice:
            logger.debug('dumping cache; now ts: %r, last_ts: %r' % (
                ts,
                self.last_timeslice)
                )
            self.cache = {}
            self.last_timeslice = ts

        result = self.cache.get(cache_key)

        return result

    def execute(self, conn, **kw):
        cache_key = (
            text_(self.base_dn % kw, 'utf-8'),
            self.scope,
            text_(self.filter_tmpl % kw, 'utf-8')
            )

        logger.debug('searching for %r' % (cache_key,))

        if self.cache_period:
            result = self.query_cache(cache_key)
            if result is not None:
                logger.debug('result for %r retrieved from cache' %
                             (cache_key,)
                             )
            else:
                result = conn.search_s(*cache_key)
                self.cache[cache_key] = result
        else:
            result = conn.search_s(*cache_key)

        logger.debug('search result: %r' % (result,))

        return result


def _timeslice(period, when=None):
    if when is None:  # pragma: no cover
        when = time.time()
    return when - (when % period)


class Connector(object):
    """ Provides API methods for accessing LDAP authentication information."""
    def __init__(self, registry, manager):
        self.registry = registry
        self.manager = manager

    def authenticate(self, login_unsafe, password_unsafe):
        """Given a login name and a password, return a tuple of ``(dn,
        attrdict)`` if the user exists and their password
        is correct.  Otherwise return ``None``.

        In a ``(dn, attrdict)`` return value, ``dn`` will be the
        distinguished name of the authenticated user.  Attrdict will be a
        dictionary mapping LDAP user attributes to sequences of values.  The
        keys and values in the dictionary values provided will be decoded
        from UTF-8, recursively, where possible.  The dictionary returned is
        a case-insensitive dictionary implemenation.

        A zero length password will always be considered invalid since it
        results in a request for "unauthenticated authentication" which should
        not be used for LDAP based authentication. See `section 5.1.2 of
        RFC-4513 <http://tools.ietf.org/html/rfc4513#section-5.1.2>`_ for a
        description of this behavior.

        If :meth:`pyramid.config.Configurator.ldap_set_login_query` was not
        called, using this function will raise an
        :exc:`pyramid.exceptions.ConfiguratorError`.

        In pyramid_ldap <= version 0.2, authenticating with a login that
        included the domain (e.g. CORP\exampleuser) would raise
        ``ldap.FILTER_ERROR`` because the ``\`` led to an invalid search
        string. In pyramid_ldap >= 0.3, the string is escaped so it will not
        raise an exception. However, it will likely fail to authenticate user
        ``CORP\\5cexampleuser`` (the escaped form of login
        "CORP\exampleuser"). Applications using pyramid_ldap can preprocess the
        logins to make sure they are formatted correctly for their
        ``ldap.login_filter_tpl`` setting.
        """
        if password_unsafe == '':
            return None

        # although we can run `search.execute(conn, login, password)` on `None`
        # values, it will come back with no results and return `None`. It's
        # better to return early here so that we know we're not passing `None`
        # values to `escape_filter_chars`, which will raise an
        # `AttributeError`.
        if login_unsafe is None or password_unsafe is None:
            return None

        # we escape untrusted inputs `login_unsafe` and `password_unsafe`
        login = escape_filter_chars(login_unsafe)
        password = escape_filter_chars(password_unsafe)

        with self.manager.connection() as conn:
            search = getattr(self.registry, 'ldap_login_query', None)
            if search is None:
                raise ConfigurationError(
                    'ldap_set_login_query was not called during setup')

            result = search.execute(conn, login=login, password=password)
            if len(result) == 1:
                login_dn = result[0][0]
            else:
                return None
        try:
            with self.manager.connection(login_dn, password) as conn:
                # must invoke the __enter__ of this thing for it to connect
                return _ldap_decode(result[0])
        except ldap.LDAPError:
            logger.debug('Exception in authenticate with login %r' % login,
                         exc_info=True)
            return None

    def user_groups(self, userdn):
        """ Given a user DN, return a sequence of LDAP attribute dictionaries
        matching the groups of which the DN is a member.  If the DN does not
        exist, return ``None``.

        In a return value ``[(dn, attrdict), ...]``, ``dn`` will be the
        distinguished name of the group.  Attrdict will be a dictionary
        mapping LDAP group attributes to sequences of values.  The keys and
        values in the dictionary values provided will be decoded from UTF-8,
        recursively, where possible.  The dictionary returned is a
        case-insensitive dictionary implemenation.

        If :meth:`pyramid.config.Configurator.ldap_set_groups_query` was not
        called, using this function will raise an
        :exc:`pyramid.exceptions.ConfiguratorError`
        """
        with self.manager.connection() as conn:
            search = getattr(self.registry, 'ldap_groups_query', None)
            if search is None:
                raise ConfigurationError(
                    'set_ldap_groups_query was not called during setup')
            try:
                result = search.execute(conn, userdn=userdn)
                return _ldap_decode(result)
            except ldap.LDAPError:
                logger.debug(
                    'Exception in user_groups with userdn %r' % userdn,
                    exc_info=True)
                return None


def ldap_set_login_query(config, base_dn, filter_tmpl,
                         scope=ldap.SCOPE_ONELEVEL, cache_period=0):
    """ Configurator method to set the LDAP login search.  ``base_dn`` is the
    DN at which to begin the search.  ``filter_tmpl`` is a string which can
    be used as an LDAP filter: it should contain the replacement value
    ``%(login)s``.  Scope is any valid LDAP scope value
    (e.g. ``ldap.SCOPE_ONELEVEL``).  ``cache_period`` is the number of seconds
    to cache login search results; if it is 0, login search results will not
    be cached.

    Example::

        config.set_ldap_login_query(
            base_dn='CN=Users,DC=example,DC=com',
            filter_tmpl='(sAMAccountName=%(login)s)',
            scope=ldap.SCOPE_ONELEVEL,
            )

    The registered search must return one and only one value to be considered
    a valid login.
    """
    query = _LDAPQuery(base_dn, filter_tmpl, scope, cache_period)

    def register():
        config.registry.ldap_login_query = query

    intr = config.introspectable(
        'pyramid_ldap login query',
        None,
        str(query),
        'pyramid_ldap login query'
        )

    config.action('ldap-set-login-query', register, introspectables=(intr,))


def ldap_set_groups_query(config, base_dn, filter_tmpl,
                          scope=ldap.SCOPE_SUBTREE, cache_period=0):
    """ Configurator method to set the LDAP groups search.  ``base_dn`` is
    the DN at which to begin the search.  ``filter_tmpl`` is a string which
    can be used as an LDAP filter: it should contain the replacement value
    ``%(userdn)s``.  Scope is any valid LDAP scope value
    (e.g. ``ldap.SCOPE_SUBTREE``).  ``cache_period`` is the number of seconds
    to cache groups search results; if it is 0, groups search results will
    not be cached.

    Example::

        config.set_ldap_groups_query(
            base_dn='CN=Users,DC=example,DC=com',
            filter_tmpl='(&(objectCategory=group)(member=%(userdn)s))'
            scope=ldap.SCOPE_SUBTREE,
            )

    """
    query = _LDAPQuery(base_dn, filter_tmpl, scope, cache_period)

    def register():
        config.registry.ldap_groups_query = query
    intr = config.introspectable(
        'pyramid_ldap groups query',
        None,
        str(query),
        'pyramid_ldap groups query'
        )
    config.action('ldap-set-groups-query', register, introspectables=(intr,))


def ldap_setup(config, uri, bind=None, passwd=None, pool_size=10, retry_max=3,
               retry_delay=.1, use_tls=False, timeout=-1, use_pool=True):
    """ Configurator method to set up an LDAP connection pool.

    - **uri**: ldap server uri **[mandatory]**
    - **bind**: default bind that will be used to bind a connector.
      **default: None**
    - **passwd**: default password that will be used to bind a connector.
      **default: None**
    - **size**: pool size. **default: 10**
    - **retry_max**: number of attempts when a server is down. **default: 3**
    - **retry_delay**: delay in seconds before a retry. **default: .1**
    - **use_tls**: activate TLS when connecting. **default: False**
    - **timeout**: connector timeout. **default: -1**
    - **use_pool**: activates the pool. If False, will recreate a connector
       each time. **default: True**
    """
    vals = dict(
        uri=uri, bind=bind, passwd=passwd, size=pool_size,
        retry_max=retry_max, retry_delay=retry_delay, use_tls=use_tls,
        timeout=timeout, use_pool=use_pool
        )

    manager = ConnectionManager(**vals)

    def get_connector(request):
        registry = request.registry
        return Connector(registry, manager)

    config.set_request_property(get_connector, 'ldap_connector', reify=True)

    intr = config.introspectable(
        'pyramid_ldap setup',
        None,
        pprint.pformat(vals),
        'pyramid_ldap setup'
        )
    config.action('ldap-setup', None, introspectables=(intr,))


def get_ldap_connector(request):
    """
    Return the LDAP connector attached to the request.  If
    :meth:`pyramid.config.Configurator.ldap_setup` was not called, using
    this function will raise an :exc:`pyramid.exceptions.ConfigurationError`.
    """
    connector = getattr(request, 'ldap_connector', None)
    if connector is None:
        raise ConfigurationError(
            'You must call Configurator.ldap_setup during setup '
            'to use an ldap connector')
    return connector


def groupfinder(userdn, request):
    """ A groupfinder implementation useful in conjunction with
    out-of-the-box Pyramid authentication policies.  It returns the DN of
    each group belonging to the user specified by ``userdn`` to as a
    principal in the list of results; if the user does not exist, it returns
    None."""
    connector = get_ldap_connector(request)
    group_list = connector.user_groups(userdn)
    if group_list is None:
        return None
    group_dns = []
    for dn, attrs in group_list:
        group_dns.append(dn)
    return group_dns


def _ldap_decode(result):
    """ Decode (recursively) strings in the result data structure to Unicode
    using the utf-8 encoding """
    return _Decoder().decode(result)


class _Decoder(object):
    """
    Stolen from django-auth-ldap.

    Encodes and decodes strings in a nested structure of lists, tuples, and
    dicts. This is helpful when interacting with the Unicode-unaware
    python-ldap.
    """

    ldap = ldap

    def __init__(self, encoding='utf-8'):
        self.encoding = encoding

    def decode(self, value):
        """
        Uses `pyramid.compat.text_` to convert values to `unicode` in python 2
        and `str` in python 3.
        """
        try:
            if isinstance(value, (bytes, str)):
                value = text_(value, encoding=self.encoding)
            elif isinstance(value, list):
                value = self._decode_list(value)
            elif isinstance(value, tuple):
                value = tuple(self._decode_list(value))
            elif isinstance(value, dict):
                value = self._decode_dict(value)
        except UnicodeDecodeError:
            pass

        return value

    def _decode_list(self, value):
        return [self.decode(v) for v in value]

    def _decode_dict(self, value):
        # Attribute dictionaries should be case-insensitive. python-ldap
        # defines this, although for some reason, it doesn't appear to use it
        # for search results.
        decoded = self.ldap.cidict.cidict()

        for k, v in iteritems_(value):
            decoded[self.decode(k)] = self.decode(v)

        return decoded


def includeme(config):
    """ Set up Configurator methods for pyramid_ldap """
    config.add_directive('ldap_setup', ldap_setup)
    config.add_directive('ldap_set_login_query', ldap_set_login_query)
    config.add_directive('ldap_set_groups_query', ldap_set_groups_query)
